infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentManager.java

```diff
@@ -0,0 +1,449 @@
+    private static final String SEGMENT_COMPACTING_FILE_SUFFIX = ".cpt";
+
+    private final StreamLogParams logParams;
+
+    private final Path logDir;
+
+    private final ResourceQuota logSizeQuota;
+
+    private final StreamLogDataStore dataStore;
+
+    private final LoadingCache<Long, StreamLogSegment> streamSegmentCache;
+
+    private final LoadingCache<Long, GarbageLogSegment> garbageSegmentCache;
+
+    // A CompactionMetadata is not removed for the whole life time of log unit.
+    private final Map<Long, CompactionMetadata> segmentCompactionMetadata = new ConcurrentHashMap<>();
+
+    SegmentManager(StreamLogParams logParams, Path logDir,
+                   ResourceQuota logSizeQuota, StreamLogDataStore dataStore) {
+        this.logParams = logParams;
+        this.logDir = logDir;
+        this.logSizeQuota = logSizeQuota;
+        this.dataStore = dataStore;
+        this.streamSegmentCache = Caffeine.newBuilder()
+                .maximumSize(logParams.maxOpenStreamSegments)
+                .removalListener(getRemovalListener())
+                .build(ordinal -> getLogSegment(ordinal, StreamLogSegment.class));
+        this.garbageSegmentCache = Caffeine.newBuilder()
+                .maximumSize(logParams.maxOpenGarbageSegments)
+                .removalListener(getRemovalListener())
+                .build(ordinal -> getLogSegment(ordinal, GarbageLogSegment.class));
+    }
+
+    private <T extends AbstractLogSegment> RemovalListener<Long, T> getRemovalListener() {
+        return (key, value, cause) -> {
+            if (value == null) {
+                return;
+            }
+            if (cause == RemovalCause.SIZE) {
+                value.close(false);
+            } else if (cause == RemovalCause.EXPLICIT) {
+                value.close(true);
+            } else {
+                log.error("RemovalListener: unexpected removal cause: {}", cause);
+            }
+        };
+    }
+
+    private <T extends AbstractLogSegment> String getSegmentFilePath(
+            long ordinal, Class<T> segmentType, boolean isCompactionOutput) {
+        String fileSuffix;
+
+        if (segmentType.equals(StreamLogSegment.class)) {
+            fileSuffix = STREAM_SEGMENT_FILE_SUFFIX;
+        } else if (segmentType.equals(GarbageLogSegment.class)) {
+            fileSuffix = GARBAGE_SEGMENT_FILE_SUFFIX;
+        } else {
+            throw new IllegalArgumentException("getSegmentFilePath: " +
+                    "Unknown segment type: " + segmentType);
+        }
+
+        if (isCompactionOutput) {
+            fileSuffix += SEGMENT_COMPACTING_FILE_SUFFIX;
+        }
+
+        return Paths.get(logDir.toString(), ordinal + fileSuffix).toString();
+    }
+
+    /**
+     * Get the ordinal of the segment that the global address belongs to.
+     *
+     * @param globalAddress the global address
+     * @return ordinal of the segment
+     */
+    long getSegmentOrdinal(long globalAddress) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651450)